File: LinkedList/SkipList/SkipList.py
import random

class SkipNode:
    def __init__(self, value, level):
        self.value = value
        self.next = [None] * (level + 1)

class SkipList:
    def __init__(self, max_level=4, p=0.5):
        self.max_level = max_level
        self.p = p
        self.header = SkipNode(float('-inf'), self.max_level)
        self.level = 0

    def random_level(self):
        lvl = 0
        while random.random() < self.p and lvl < self.max_level:
            lvl += 1
        return lvl

    def insert(self, value):
        update = [None] * (self.max_level + 1)
        current = self.header

        for i in range(self.level, -1, -1):
            while current.next[i] and current.next[i].value < value:
                current = current.next[i]
            update[i] = current

        current = current.next[0]

        if current is None or current.value != value:
            rlevel = self.random_level()

            if rlevel > self.level:
                for i in range(self.level + 1, rlevel + 1):
                    update[i] = self.header
                self.level = rlevel

            new_node = SkipNode(value, rlevel)

            for i in range(rlevel + 1):
                new_node.next[i] = update[i].next[i]
                update[i].next[i] = new_node

    def append(self, value):
        self.insert(value)

    def prepend(self, value):
        self.insert(value)

    def delete(self, value):
        update = [None] * (self.max_level + 1)
        current = self.header

        for i in range(self.level, -1, -1):
            while current.next[i] and current.next[i].value < value:
                current = current.next[i]
            update[i] = current

        current = current.next[0]

        if current and current.value == value:
            for i in range(self.level + 1):
                if update[i].next[i] != current:
                    break
                update[i].next[i] = current.next[i]

            while self.level > 0 and self.header.next[self.level] is None:
                self.level -= 1

    def search(self, value):
        current = self.header
        for i in range(self.level, -1, -1):
            while current.next[i] and current.next[i].value < value:
                current = current.next[i]
        current = current.next[0]
        return current is not None and current.value == value

    def print(self):
        print("\n--- Skip List Levels ---")
        for lvl in range(self.level, -1, -1):
            print(f"Level {lvl}: ", end="")
            node = self.header.next[lvl]
            while node:
                print(f"{node.value} -> ", end="")
                node = node.next[lvl]
            print("None")

if __name__ == "__main__":
    sl = SkipList()
    sl.insert(3)
    sl.insert(6)
    sl.insert(7)
    sl.insert(9)
    sl.insert(12)
    sl.print()

    print(f"Searching for 7: {sl.search(7)}")
    sl.delete(7)
    sl.print()

File: LinkedList/SkipList/SkipList.java
import java.util.Arrays;
import java.util.Random;

public class SkipList {
    private static class SkipNode {
        int value;
        SkipNode[] next;

        SkipNode(int value, int level) {
            this.value = value;
            this.next = new SkipNode[level + 1];
        }
    }

    private final int MAX_LEVEL = 4;
    private final float P = 0.5f;
    private int level = 0;
    private final SkipNode header;
    private final Random rand = new Random();

    public SkipList() {
        header = new SkipNode(Integer.MIN_VALUE, MAX_LEVEL);
    }

    private int randomLevel() {
        int lvl = 0;
        while (rand.nextFloat() < P && lvl < MAX_LEVEL) {
            lvl++;
        }
        return lvl;
    }

    public void insert(int value) {
        SkipNode[] update = new SkipNode[MAX_LEVEL + 1];
        SkipNode current = header;

        for (int i = level; i >= 0; i--) {
            while (current.next[i] != null && current.next[i].value < value) {
                current = current.next[i];
            }
            update[i] = current;
        }
        current = current.next[0];

        if (current == null || current.value != value) {
            int rlevel = randomLevel();

            if (rlevel > level) {
                for (int i = level + 1; i <= rlevel; i++) {
                    update[i] = header;
                }
                level = rlevel;
            }

            SkipNode newNode = new SkipNode(value, rlevel);
            for (int i = 0; i <= rlevel; i++) {
                newNode.next[i] = update[i].next[i];
                update[i].next[i] = newNode;
            }
        }
    }

    public boolean search(int value) {
        SkipNode current = header;
        for (int i = level; i >= 0; i--) {
            while (current.next[i] != null && current.next[i].value < value) {
                current = current.next[i];
            }
        }
        current = current.next[0];
        return current != null && current.value == value;
    }

    public void delete(int value) {
        SkipNode[] update = new SkipNode[MAX_LEVEL + 1];
        SkipNode current = header;

        for (int i = level; i >= 0; i--) {
            while (current.next[i] != null && current.next[i].value < value) {
                current = current.next[i];
            }
            update[i] = current;
        }
        current = current.next[0];

        if (current != null && current.value == value) {
            for (int i = 0; i <= level; i++) {
                if (update[i].next[i] != current) break;
                update[i].next[i] = current.next[i];
            }
            while (level > 0 && header.next[level] == null) {
                level--;
            }
        }
    }

    public void print() {
        System.out.println("\n--- Skip List Levels ---");
        for (int i = level; i >= 0; i--) {
            SkipNode node = header.next[i];
            System.out.print("Level " + i + ": ");
            while (node != null) {
                System.out.print(node.value + " -> ");
                node = node.next[i];
            }
            System.out.println("null");
        }
    }

    public static void main(String[] args) {
        SkipList sl = new SkipList();
        sl.insert(4);
        sl.insert(8);
        sl.insert(15);
        sl.print();
        System.out.println("Search 8: " + sl.search(8));
        sl.delete(8);
        sl.print();
    }
}
